### PR TITLE
feat: bulk writer 555 rate_limiter

### PR DESCRIPTION
**Pull request by craiglabenz**
Destined for the `bulk-writer-release` branch, where it will meet up with its friends that collectively comprise the bulk-writer feature.

### CODE REVIEW

File: google/cloud/firestore_v1/throttle.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+class Throttle:
+    """Implements 5/5/5 ramp-up via Token Bucket algorithm."""
+
+    def __init__(
+        self,
+        initial_tokens: Optional[int] = default_initial_tokens,
+        phase_length: Optional[int] = default_phase_length,
+        now_getter: Optional[Callable] = datetime.datetime.utcnow,
+    ):
+        # Tracks the volume of operations during the
```
Comment (tseaver) [inline google/cloud/firestore_v1/throttle.py:35]: Incomplete sentence.

File: google/cloud/firestore_v1/throttle.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        self._operations_this_phase: int = 0
+
+        # Optional override for telling time. Useful for testing.
+        self._now = now_getter
```
Comment (tseaver) [inline google/cloud/firestore_v1/throttle.py:39]: Is using `mock.patch(`datetime.datetime.utcnow`), either within a testcase or as a wrapper around it, not a clearer pattern?  We've removed a bunch of `_now` thingies from other code over the years.  E.g., [here](https://github.com/googleapis/python-storage/blob/2c87f2f1438150bd93c7d544d2ccd4fdf95998db/tests/unit/test__signing.py#L96-L108).

File: google/cloud/firestore_v1/throttle.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        operations_last_phase: int = self._operations_this_phase
+        self._operations_this_phase = 0
+
+        previous_phase: int = self._phase
+        self._phase = expected_phase
+
+        # No-op if we did nothing for an entire phase
+        if not operations_last_phase or self._phase > previous_phase + 1:
+            return
+        self._increase_maximum_tokens()
```
Comment (tseaver) [inline google/cloud/firestore_v1/throttle.py:98]: Might be clearer as:
```suggestion
        if operations_last_phase and self._phase <= previous_phase + 1:
            self._increase_maximum_tokens()
```

**Review by tseaver — COMMENTED** [review]
Is this PR complete?  The title is misleading, if so.

File: google/cloud/firestore_v1/throttle.py (hunk 2 above)
Comment (craiglabenz) [reply google/cloud/firestore_v1/throttle.py:39]: Hmm, potentially. My fear was that, as this is an extremely clock-sensitive feature, all the mocking and re-mocking to set the clock as just-so would end up being more tedious over all. But, if the removal of fake `_now`-type objects has already been a standing effort, that could swing the balance.

File: google/cloud/firestore_v1/throttle.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    return datetime.datetime.utcnow()
+
+
+default_initial_tokens: int = 500
+default_phase_length: int = 60 * 5  # 5 minutes
+microseconds_per_second: int = 1000000
+
+
+class Throttle:
+    """Implements 5/5/5 ramp-up via Token Bucket algorithm."""
```
Comment (crwilcox) [inline google/cloud/firestore_v1/throttle.py:29]: could this comment be expanded a bit to explain what is going on here?

File: google/cloud/firestore_v1/throttle.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    def __init__(
```
Comment (crwilcox) [inline google/cloud/firestore_v1/throttle.py:31]: can you add a docstring here describing using these inputs?

File: google/cloud/firestore_v1/throttle.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        self._maximum_tokens: int = self._available_tokens
+
+        # Number of seconds after which the [_maximum_tokens] can increase by 50%.
+        self._phase_length: int = phase_length
+
+        # Tracks how many times the [_maximum_tokens] has increased by 50%.
+        self._phase: int = 0
+
+    def take_tokens(self, num: Optional[int] = 1) -> bool:
+        """Returns True if an operation is currently permitted or False if not."""
```
Comment (crwilcox) [inline google/cloud/firestore_v1/throttle.py:57]: what does it mean to be permitted?

Comment (craiglabenz) [reply google/cloud/firestore_v1/throttle.py:57]: this docstring and type annotation is actually just outright stale. updating.

File: google/cloud/firestore_v1/throttle.py (hunk 2 above)
Comment (craiglabenz) [reply google/cloud/firestore_v1/throttle.py]: Made this change.

**Comment by tseaver** [comment]
@googlebot I consent.

**Comment by tseaver** [comment]
@craiglabenz The CLA bot sometimes gets confused about the CLAs of through-the-web suggestion authors.

**Comment by craiglabenz** [comment]
@tseaver So I've heard, haha. In my opinion, this PR is ready to land in the release branch. Feel free to LGTM if you agree :)

**Review by tseaver — APPROVED** [review]

**Pull request merged**